### DISCUSSION TITLE
Generic AutoDataAttribute

### DISCUSSION THREAD

**Discussion by isberg** [AutoFixture/AutoFixture#1302] [Ideas]
Some time ago I got the idea to pass my custom fixture type as a generic parameter to AutoDataAttribute, then I got sad because it was impossible (because of .NET) and then happy when I realised it will be possible in .NET 6. I am happy to say that it is now possible (using .NET 6 and LangVersion=preview). 

```
[AttributeUsage(AttributeTargets.All,AllowMultiple=true)]
internal class AutoDataAttribute<TFixture> : AutoDataAttribute where TFixture : IFixture
{
	public AutoDataAttribute() : base(() => Activator.CreateInstance<TFixture>()) { }
}
```

It can then be used like `[Theory, AutoData<OneFixture>, AutoData<OtherFixture>]`. I think I also managed to get it to work with InlineAutoData, but when going for `AutoMockDataAttribute<TFixture>` it just gave me a headache.

So, I just want to share that this is now (tentatively?) possible and wonder if it is something that would be useful to include out of the box.

**Comment by aivascu** [comment]
Interesting, this might be worth looking into. Perhaps the generic argument should not be an `IFixture` type but rather something that knows how to create an `IFixture`.  For example some `IFixtureFactory`, where `IFixtureFactory` would have a member `Factory` that is a `Func<IFixture>`.

```cs
class AutoDataAttribute<T> : AutoDataAttribute where T: IFixtureFactory, new()
{
    public AutoDataAttribute() : base(new T().Factory) { }
}
```
If this syntax was possible it would solve the issue of having to create multiple variations of custom attributes for every minor difference in the fixture customization. It would move the problem elsewhere though, but it wouldn't be the attributes.

**Reply by isberg** [reply]
I will try out the variant with a fixture factory on my sample code, but I like it already. Feels more powerful.

**Reply by isberg** [reply]
I have been thinking and trying out some code using both an IFixture and an IFixtureFactory as a generic parameter and I am not sure I see the benefit of creating a fixture factory implementation for every fixture I want to pass to the generic attribute. I might be missing something though. I think something like this would work pretty smooth. 

``` csharp
[AttributeUsage(AttributeTargets.All, AllowMultiple = true)]
public class AutoDataAttribute<T> : AutoDataAttribute where T: IFixture, new()
{
    public AutoDataAttribute() : base(() => new T()) { }
}

public class InlineAutoDataAttribute<T> : InlineAutoDataAttribute where T : IFixture, new()
{
    public InlineAutoDataAttribute(params object[] values) : base(new AutoDataAttribute<T>(), values) { }
}
```

**Reply by Kralizek** [reply]
How about this?

```csharp
public interface IFixtureCustomizer
{
    void CustomizeFixture(IFixture fixture);
}

[AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
public class GenericAutoDataAttribute<T> : AutoDataAttribute
    where T : IFixtureCustomizer, new()
{
    public GenericAutoDataAttribute() : base (() => CreateFixture()) {}
    
    private static IFixture CreateFixture()
    {
        var fixture = new Fixture();
        
        var customizer = new T();
        
        customizer.CustomizeFixture(fixture);
        
        return fixture;
    }
}
```

**Reply by isberg** [reply]
@Kralizek I guess the benefit of this solution over the previous two would be that the user only nees to create one type and the ability to compose multiple customizers. I think I like it. :-)

**Reply by Kralizek** [reply]
Also it's easy to wrap these two types in compiler directives without polluting existing types.

```csharp
#if NET6_0_OR_GREATER

#endif
```

Source: https://docs.microsoft.com/en-us/dotnet/standard/frameworks#how-to-specify-a-target-framework

**Reply by isberg** [reply]
How about instead use the already existing ICustomization?

``` csharp
[AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
public class GenericAutoDataAttribute<T> : AutoDataAttribute
    where T : ICustomization, new()
{
    public GenericAutoDataAttribute() : base(() => CreateFixture()) { }

    private static IFixture CreateFixture()
    {
        return new Fixture().Customize(new T());
    }
}
```

The above leads me to a second extended version like:

``` csharp
[AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
public class GenericAutoDataAttribute<TFixture, TCustomization> : AutoDataAttribute
    where TCustomization : ICustomization, new()
    where TFixture : IFixture, new()
{
    public GenericAutoDataAttribute() : base(() => CreateFixture()) { }

    private static IFixture CreateFixture()
    {
    	return new TFixture().Customize(new TCustomization());
    }
}
```

**Reply by Kralizek** [reply]
Uhm... never seen a need for providing my own implementation of `IFixture` or a specialization of `Fixture`, do you have a case for it?

Regarding reusing `ICustomization` or introducing a new type, I guess it's about API clarity vs reusability.

**Reply by aivascu** [reply]
@Kralizek AutoFixture always supported users in creating their own entry points for the library. The original maintainers always pointed out that the `Fixture` class is just the entry point provided by the library. In this context `IFixture` is just a spec for implementing a custom Fixture that might have its own default customizations, residue collectors or even engine parts. I believe it is important that we offer the same amount of flexibility going forward.

One use case in favor of not using a specific implementation of `IFixture` is the necessity to test the library itself. You'll find on many examples in the AutoFixture code base where `IFixture` stubs were used.

Also there's a saying "[new is glue](https://ardalis.com/new-is-glue/)" 🙂.

**Reply by isberg** [reply]
@Kralizek when I am creating my own fixture class it is normally just extending Fixture and adding a bit of customization in the constructor. Often to use when extending /Inline/AutoDataAttribute, but sometimes I just write it inline in the constructor of AutoDataAttribute, if I don't need both versions of the attribute. 

Regarding reusing ICustomization, considering that the documentation says "Encapsulates a customization of an IFixture", I think it would make quite alot of sense being able to say something like `[Theory,
AutoData<AutoMoqCustomization>]`.

I guess the question is what is least confusing to most users. 

@aivascu I guess I should see if I can write tests for these attributes. :-)

**Reply by aivascu** [reply]
@isberg tests are the thing you should have started with, especially considering the kind of library this is. 🙂

I don't think the generic parameter can be an `ICustomization` since this would imply you're making a decision on what the `IFixture` implementation is. Also I don't think it should be an `IFixture` since that interface is not very trivial to implement, and might lead people to make bad decisions. This is why I proposed a new interface that would know how to create the `IFixture` but wouldn't require a lot of ceremony to create.

A benefit of this approach is that the factory could also contain all the necessary customizations for the test, and the name would make them explicit. Consider the following implementation.

```cs
public class CustomizingFactory : IFixtureFactory
{
    private readonly Func<IFixture> factory;
    private readonly ICustomization customization;

    public CustomizingFactory(ICustomization customization)
        : this(() => new Fixture(), customization)
    {
    }

    public CustomizingFactory(Func<IFixture> factory, ICustomization customization)
    {
        this.factory = factory;
        this.customization = customization;
    }

    public Func<IFixture> Factory => () => this.factory().Customize(this.customization);
}

public class BillingData : CustomizingFactory
{
    public BillingData()
        : base(new CompositeCustomization(
            new AutoMoqCustomization(),
            new BillingCustomization()))
    {
    }
}

[Theory]
[AutoData<BillingData>]
public void Foo(Mock<IExchange> exchangeMock, TaxCalculator calculator)
{
    // ...
}
```
I would prefer not to introduce a new interface, and especially one that has such a limited use, so let's see if we can find a better way.

**Reply by Kralizek** [reply]
@aivascu how would you use your proposed classes/interfaces to add concrete customizations?

In my personal experience, the need for multiple attributes comes from the fact that in some tests you might want to include or omit a certain property.

I'd appreciate if this conversation removed this great pain point. :)

**Reply by aivascu** [reply]
@Kralizek properties? You mean like class properties? Could you provide an example? This is not something I would expect custom attributes to be created for.

I would expect that a specific value would be created in the test itself.

In my example customizations  would be used by composing them.

**Reply by isberg** [reply]
@Kralizek I wonder if your scenario is the same as mine, which I will try to express with an example.

Consider that I have an `Order` which have a default tax percentage and no discount, and I want to be able to direct autofixture when to generate different tax and discount, and possibly both, but as a default all properties should be left alone.

I am able to customize this nicely in one swoop, but when I try to split it, it does not seem to combine, but only one of the customizations will be applied.

``` csharp
fixture
    .Customize<Order>(c => c
        .OmitAutoProperties()
        .With<decimal, int>(x => x.DiscountPercent, y => y % 20 * 5)
        .With<decimal, int>(x => x.TaxPercent, y => y % 8 * 5)
    );
```

I thought that maybe I could do `[Theory, AutoData<TaxFree, Discount>]` and combine two customizations of the same type:

``` csharp
class TaxFree : ICustomization { }
class Discount : ICustomization { }
```

I also though that maybe use a custom AutoDataAttribute with the default configuration (OmitAutoProperties in this case) and then pass a customization as the generic parameter (`[Theory, AutoOrder, AutoOrder<TaxFree>, AutoOrder<Discount>, AutoOrder<TaxFree, Discount>]`), but that I assume would in same way as the above example. 

@aivascu do you know if there is any way to apply the above `Customize` in multiple calls? Or if it is possible to do it smoothly in a completely different way out of the box?

For completeness the Order looks like this:

``` csharp
record Order(decimal Subtotal)
{
    public decimal TaxPercent { get; init; } = 25m;
    public decimal DiscountPercent { get; init; } = 0m;
	
    public decimal Discount => DiscountPercent / 100m * Subtotal;
    public decimal Tax => TaxPercent / 100m * (Subtotal - Discount);
	
    public decimal Total => Subtotal - Discount + Tax;
}
```

**Reply by Kralizek** [reply]
@isberg Yup, it looks very similar to my issue.

This is the setup I have for one of my projects:

```csharp
[AttributeUsage(AttributeTargets.Method)]
public class CustomAutoDataAttribute : AutoDataAttribute
{
    public CustomAutoDataAttribute() : base(FixtureHelpers.CreateFixture)
    {
    }
}

[AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
public class CustomInlineAutoDataAttribute : InlineAutoDataAttribute
{
    public CustomInlineAutoDataAttribute(params object[] args) : base(FixtureHelpers.CreateFixture, args)
    {
    }
}

internal static class FixtureHelpers
{
    public static IFixture CreateFixture()
    {
        var fixture = new Fixture();

        fixture.Customize(new AutoMoqCustomization
        {
            ConfigureMembers = true,
            GenerateDelegates = true
        });

        fixture.AddGrpcSupport();

        fixture.Customizations.Add(new MockRelay(new ExactTypeSpecification(typeof(CustomerResourceAccessClient))));

        fixture.Customizations.Add(new MockRelay(new ExactTypeSpecification(typeof(ProviderResourceAccessClient))));

        fixture.Customizations.Add(new TypeRelay(typeof(ILogger<>), typeof(TestContextLoggerAdapter<>)));

        fixture.Customize<PatientModel>(o => o.With(p => p.Email, (MailAddress mail) => mail.Address).With(p => p.PersonalIdentityNumber, (SwedishPersonalIdentityNumber ssn) => ssn.ToFormattedString(SwedishPersonalIdentityNumberFormats.TwelveDigits)));

        return fixture;
    }
}
```

**Reply by aivascu** [reply]
@isberg by default AutoFixture will apply only the last `Customize<T>` expression because it generates a new customization which is placed on top of all other customizations and it will be the first one to handle the request for `T`.

There are multiple ways you could compose customizations for a type in AutoFixture.

If you want to compose parts of `Customize<T>(c => c)` expressions you have to work with the `IPostprocessComposer<T>`. There's an example of how I see the code reuse in the context of those abstractions, in this SO [answer](https://stackoverflow.com/a/66093487/2542567). I created a [gist](https://gist.github.com/aivascu/b9c5d76b501e2bc1b53af9aae6a61746) to show how I would attempt to do composition of prostprocessors, if I had to.

Using the `ICustomization` you could achieve the same effect but it wouldn't be as pretty, though it might perform faster in some scenarios. You could abstract away nested instantiations with an extension method. See this [example](https://gist.github.com/aivascu/6a5da6288940dfadb9cbaead610df3c9).

A different approach would be to create custom builders using `ISpecimenBuilder` and using them with `FromFactory(ISpecimenBuilder)`. There you could take full ownership of creating instance prototypes using AutoFixture tooling and reflection. It would be possible to further customize these instances using the other postprocessor actions like `With`, `Without`, etc.

Another approach involving `ISpecimenBuilder` would be, to create builders that reply to specific request objects, like for example `DiscountedProductOrderRequest`. These would work similar to the `SeededRequest` and would contain all the necessary data to create the instances. This solution could be improved further with the implementation of `IParameterCustomizationSource` attributes (similar to `[Frozen]`), that would relay the object creation only for specific test parameters.

All approaches have benefits and drawbacks. 

I personally prefer to create customizations for each area of my application, (e.g. `WebApiCustomization`, `BillingCustomization`, `PersistenceCustomization`), which are composed of more specialized customizations like `AutoMoqCustomization`, `BillingDomainData`, `EntityFrameworkCustomization`, etc. These higher level customizations are used via data atrribute `WebApiDataAttribute`, `InlineBillingDataAttribute`, etc. 
For parts that differ from the main behavior of customizations I create the data inside the test using `fixutre.Build<T>()`, and inject it wherever it is appropriate.

**Reply by aivascu** [reply]
@Kralizek does the above reply help you?

**Reply by Kralizek** [reply]
In a way, but I don't see often the need to create too many attributes. This is why I'd prefer some low ceremony interface to implement.

**Reply by aivascu** [reply]
@Kralizek I agree that whatever is the generic parameter, it should take care of the `Func<IFixture>` part of the original attributes. All other issues can be solved with existing features of AutoFixture.

**Comment by isberg** [comment]
@Kralizek does it mean then that you would happy to know that I was able to get the following to work:

``` csharp
[Theory, Order<Discount, TaxFree>]
void DiscountTaxFree(Order sut)
{
	Assert.Equal(0, sut.Tax);
	Assert.Equal(sut.Total, sut.Subtotal - sut.Discount);
}
```

It is not very beutiful (yet) though:

``` csharp
public class OrderAttribute<T1, T2> : AutoDataAttribute
	where T1 : IPostprocessingStep<Order>, new()
	where T2 : IPostprocessingStep<Order>, new()
{
	public OrderAttribute() : base(CreateFixture) { }

	static IFixture CreateFixture()
	{
		var result = new Fixture();
		result
			.Customize<Order>(c => c
				.OmitAutoProperties()
				.Process(new T1())
				.Process(new T2()));
		return result;		
	}
}

public static class PostprocessComposerExtensions
{
	public static IPostprocessComposer<T> Process<T>(this IPostprocessComposer<T> self, IPostprocessingStep<T> ps)
		=> ps.Process(self);
}

public interface IPostprocessingStep<T>
{
	IPostprocessComposer<T> Process(IPostprocessComposer<T> composer);
}

public class Discount : IPostprocessingStep<Order>
{
	public IPostprocessComposer<Order> Process(IPostprocessComposer<Order> composer)
		=> composer.With<decimal, int>(p => p.DiscountPercent, y => y % 20 * 5);
}
```

There are also two more versions of OrderAttribute with 1 and 0 generic paramters and lots of duplication as well as a second IPostprocessingStep<Order>. I hope there is a bit more that can be done to generalise it, but likely not today.

@aivascu it is pretty clear that I still not found any convincing need for a generic AutoDataAttribute or what type the generic attribute should be. I have not given up yet though. :-)

**Reply by aivascu** [reply]
@isberg I think that there is a convincing need for generic auto-data attributes, just not scenarios that complicated.
The generic attributes will save developers from having to write custom attributes for each attribute type in the library (classic auto-data, inline, and upcoming member and class attributes), especially since in between attributes the only part that changed is the `Func<IFixture>` constructor parameter.
Instead they would be able to reuse composed fixture factories. Not sure if the tests would look prettier, but it would be less code for sure.

**Reply by isberg** [reply]
@aivascu I think the combination of fixture factories with IPartialPostprocessComposer has great potential, thanks for the code example. I will fiddle with the combination of your original IFixtureFactory and your IPartialPostprocessComposer.

**Reply by isberg** [reply]
From my fiddling I was able to tinker together the following code (reduntant and previously mentioned parts excluded):

``` csharp
[Theory]
[AutoData]
[AutoData<OrderFactory>]
[AutoData<OrderFactory<TaxFree>>]
[AutoData<OrderFactory<Discount>>]
[AutoData<OrderFactory<Tax>>]
[AutoData<OrderFactory<TaxFree, Discount>>]
void Test(Order sut)
{
	sut.TaxPercent.Should().BeInRange(0, 100);
	sut.DiscountPercent.Should().BeInRange(0, 50);
}

class Discount : IPartialPostprocessComposer<Order>
{
	public IPostprocessComposer<Order> Compose(IPostprocessComposer<Order> composer)
		=> composer.With(x => x.DiscountPercent, (int x) => (x % 10 + 1) * 5);
}

class OrderFactory<T> : IFixtureFactory
	where T : IPartialPostprocessComposer<Order>, new()
{
	public Func<IFixture> Factory => CreateFixture;
	IFixture CreateFixture()
	{
		var fixture = new Fixture();
		fixture
			.Customize<Order>(c => c
				.OmitAutoProperties()
				.Using(new T()));
		return fixture;
	}
}
```

AutoData just supports a generic parameter of IFixtureFactory and then it is possible to extend on this, especially if you got IPartialPostprocessComposer<T> with the supporting classes. :-) I think still OrderFactory is a bit ugly, but that is no fault of AutoFixture.

**Reply by aivascu** [reply]
@isberg as much as you try I don't think you'll be able to provide all the customizations as generic parameters for the attribute. The language simply does not allow providing a collection of generic parameters. For each added parameter you'll have to add a new class.
Even if it did, imagine the complexity when you start to factor in multiple test parameters and all the customizations you might want to apply to each of those. Add to that the complexity if you'd want to use an `InlineAutoData` attribute and also provide inline arguments.

Unfortunately there is no better way than to compose the object creation strategy, either with builders or customizations.

As mentioned in the previous reply you could apply a customization more locally by using `IParameterCustomizationSource` attributes. Here's an example of the syntax, you can check out the full example [here](https://gist.github.com/aivascu/f7567bec83fd0dd79c12d0a7871abc92).

```cs
[Theory, AutoData]
public void CustomizesDecoratedOrder([BlackFridayOrder(discount: 20d)] Order blackFridayOrder, Order order)
{
    Assert.Equal(0, blackFridayOrder.Tax);
    Assert.Equal(20, blackFridayOrder.Discount);

    Assert.NotEqual(0, order.Tax);
    Assert.NotEqual(20, order.Discount);
}
```

**Reply by isberg** [reply]
Thanks @aivascu!
